### PR TITLE
feat(package-manager): prev_graph diff from current lockfile (#438 slice 4d)

### PR DESCRIPTION
**Pull request by zkochan**
## Summary

Sub-slice 4d of #438. Closes out Slice 4 by adding the `prev_graph` diff — the input Slice 5's linker will subtract from `graph` to identify orphaned directories that need removing.

## What lands

`lockfile_to_hoisted_dep_graph` now takes an optional `current_lockfile: Option<&Lockfile>` and runs a second walk over it (when present and non-empty) with `force: true, skipped: BTreeSet::new()` to populate `result.prev_graph`. Ports upstream's wrapper at [installing/deps-restorer/src/lockfileToHoistedDepGraph.ts:70-79](https://github.com/pnpm/pnpm/blob/94240bc046/installing/deps-restorer/src/lockfileToHoistedDepGraph.ts#L70-L79).

```rust
let prev_graph = match current_lockfile {
    Some(current) if current.packages.is_some() => {
        let prev_opts = LockfileToHoistedDepGraphOptions {
            force: true,
            skipped: BTreeSet::new(),
            ..opts.clone()
        };
        Some(build_dep_graph(current, &prev_opts)?.graph)
    }
    _ => None,
};
```

## Why force + empty skipped

Both overrides matter — neither is paranoia:

- **`force: true`** — an orphan that landed under the previous install but would now fail installability (host platform changed, dependency was downgraded out of engine range, etc.) still has a directory on disk. Without `force`, the prev-walk would silently filter it out and the linker would leave the stale dir in place. Pinned by `prev_graph_includes_orphan_even_when_now_incompatible`.
- **`skipped: empty`** — the previous install's filter set says which packages it *didn't* install, which is unrelated to which directories *currently exist*. Inheriting the wanted-walk's skipped would underreport `prev_graph` for unchanged-but-now-incompatible packages.

## API change

`lockfile_to_hoisted_dep_graph` gains a middle `current_lockfile: Option<&Lockfile>` argument. The previous body splits into a private `build_dep_graph` helper that the public wrapper calls once or twice depending on whether a current lockfile is present. Existing tests updated (the only callers — `None` for the no-prev case).

## `prev_graph` shape

| `current_lockfile` | `result.prev_graph` |
|---|---|
| `None` | `None` |
| `Some(lf)` with `lf.packages: None` | `None` |
| `Some(lf)` with `lf.packages: Some(_)` | `Some(graph)` |

Upstream uses `prevGraph = {}` for both null-current cases; pacquet uses `None`. The linker treats them the same (no orphans to remove on a fresh install).

## Tests

Four new prev_graph tests; the 15 walker tests from 4a-4c carry over with the new signature.

- `prev_graph_none_when_current_lockfile_absent` — no current lockfile → `None`, wanted graph still produced.
- `prev_graph_none_when_current_lockfile_has_no_packages` — current lockfile with `packages: None` → `None`.
- `prev_graph_contains_orphan_from_current_only_lockfile` — package in current but not wanted appears in `prev_graph`, not in `graph`.
- `prev_graph_includes_orphan_even_when_now_incompatible` — darwin-only orphan in current, host is linux, wanted is empty: `prev_graph` still contains it (proves `force: true` works), and the wanted-walk's `skipped` stays empty (proves the prev-walk's skipped set is independent).

## Test plan

- [x] All 19 walker tests pass (15 carry-over + 4 new).
- [x] `just ready` (874 → 878 tests, all green).
- [x] `cargo doc --document-private-items`, `taplo format --check`, `just dylint` clean.
- [x] Sabotage check: changing the prev-walk's options from `force: true, skipped: empty` to `..opts.clone()` re-fails `prev_graph_includes_orphan_even_when_now_incompatible` (orphan disappears from `prev_graph` because it would now fail installability).

## What's next

Slice 4 is done. Next on the critical path: **Slice 5 — `linkHoistedModules`**. Consumes `LockfileToDepGraphResult` to produce the on-disk tree: walks `prev_graph` minus `graph` to rimraf orphans, then imports each graph node from the CAS via Slice 1's `import_indexed_dir` primitive, then runs `linkBins` per `node_modules/` directory.

---
Written by an agent (Claude Code, claude-opus-4-7).

## Summary by CodeRabbit

* **Improvements**
  * Added support for computing a previous dependency graph when a current lockfile is provided, improving detection of changes between lockfile states and handling of current-only (orphan) packages.

* **Tests**
  * Expanded tests for lockfile-comparison behavior, ensuring prev-graph semantics and orphaned package detection (including installability edge cases).

[Review Change Stack](https://app.coderabbit.ai/change-stack/pnpm/pacquet/pull/494)

### CODE REVIEW

File: crates/package-manager/src/hoisted_dep_graph.rs [hunk 1]
```diff
@@ -271,20 +276,59 @@ pub enum HoistedDepGraphError {
...
+    let prev_graph = match current_lockfile {
+        Some(current) if current.packages.is_some() => {
+            let prev_opts = LockfileToHoistedDepGraphOptions {
+                force: true,
+                skipped: BTreeSet::new(),
+                ..opts.clone()
+            };
+            Some(build_dep_graph(current, &prev_opts)?.graph)
+        }
+        _ => None,
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/package-manager/src/hoisted_dep_graph.rs:326]: Good catch — fixed in 84fb24d8c6.

Tightened the guard to `is_some_and(|packages| !packages.is_empty())`. The doc-comment said `None` was the contract for 'no current packages', but `is_some()` also matched `Some(empty_map)`, leaving the empty case as `Some(empty)` — inconsistent with the no-packages case which was `None`. Both cases produce the same observable "no orphans to consider" behavior, so pacquet collapses them into one representation (`None`), which also skips the no-op second walk.

Upstream uses an always-present `DependenciesGraph` (`{}` for the no-current case), so the JS-side `packages != null` guard is enough there. Pacquet uses `Option<DependenciesGraph>` and benefits from the stricter guard.

Added `prev_graph_none_when_current_lockfile_has_empty_packages` to pin the empty-map case. 20 walker tests now pass.

---
Written by an agent (Claude Code, claude-opus-4-7).

**Comment by zkochan** [comment]
@coderabbitai full review
